src/shared/keyboard-navigation.js

```diff
@@ -0,0 +1,197 @@
+import { useEffect } from 'preact/hooks';
+
+import { ListenerCollection } from './listener-collection';
+
+/**
+ * @param {HTMLElement & { disabled?: boolean }} element
+ */
+function isElementDisabled(element) {
+  return typeof element.disabled === 'boolean' && element.disabled;
+}
+
+/** @param {HTMLElement} element */
+function isElementVisible(element) {
+  return element.offsetParent !== null;
+}
+
+/**
+ * Enable arrow key navigation between interactive descendants of a
+ * container element.
+ *
+ * In addition to moving focus between elements when arrow keys are pressed,
+ * this also implements the "roving tabindex" pattern which sets the `tabindex`
+ * attribute of elements to control which element gets focus when the user
+ * tabs into the container.
+ *
+ * See [1] for a reference of how keyboard navigation should work in web
+ * applications and how it applies to various common widgets.
+ *
+ * @example
+ *   function MyToolbar() {
+ *     const container = useRef();
+ *
+ *     // Enable arrow key navigation between interactive elements in the
+ *     // toolbar container.
+ *     useArrowKeyNavigation(container);
+ *
+ *     return (
+ *       <div ref={container} role="toolbar">
+ *         <button>Bold</bold>
+ *         <button>Italic</bold>
+ *         <a href="https://example.com/help">Help</a>
+ *       </div>
+ *     )
+ *   }
+ *
+ * [1] https://www.w3.org/TR/wai-aria-practices/#keyboard
+ *
+ * @param {import('preact').RefObject<HTMLElement>} containerRef
+ * @param {object} options
+ *   @param {boolean} [options.autofocus] - Whether to focus the first element
+ *     in the set of matching elements when the component is mounted
+ *   @param {boolean} [options.horizontal] - Enable navigating elements using left/right arrow keys
+ *   @param {boolean} [options.vertical] - Enable navigating elements using up/down arrow keys
+ *   @param {string} [options.selector] - CSS selector which specifies the
+ *     elements that navigation moves between
+ */
+export function useArrowKeyNavigation(
+  containerRef,
+  {
+    autofocus = false,
+    horizontal = true,
+    vertical = true,
+    selector = 'a,button',
+  } = {}
+) {
+  useEffect(() => {
+    if (!containerRef.current) {
+      throw new Error('Container ref not set');
+    }
+    const container = containerRef.current;
+
+    const getNavigableElements = () => {
+      const elements = /** @type {HTMLElement[]} */ (
+        Array.from(container.querySelectorAll(selector))
+      );
+      return elements.filter(
+        el => isElementVisible(el) && !isElementDisabled(el)
+      );
```

Comment:
Indeed this won't catch all possible changes that could make an element invisible. In all the use cases I've looked at so far this didn't matter because when when elements appeared or disappeared it was because they were being rendered or not-rendered by a component, so catching node additions/removals was enough. We could potentially observe the `class` and `style` attributes of elements in the container as these will most likely change if eg. `display: none` or `visibility: hidden` was set via either an inline style or a class.